src/wh_client.c

```diff
@@ -201,3 +203,137 @@ int wh_Client_Echo(whClientContext* c, uint16_t snd_len, const void* snd_data,
+    return wh_Client_SendRequest(c, WH_MESSAGE_GROUP_KEY, WH_KEY_CACHE,
+            WOLFHSM_PACKET_STUB_SIZE + sizeof(packet->keyCacheReq) + inSz,
+            (uint8_t*)packet);
+}
+
+int wh_Client_KeyCacheRequest(whClientContext* c, uint32_t flags,
+    uint8_t* label, uint8_t* in, uint32_t inSz)
+{
+    return wh_Client_KeyCacheRequest_ex(c, flags, label, in, inSz,
+        WOLFHSM_ID_ERASED);
+}
+
+int wh_Client_KeyCacheResponse(whClientContext* c, uint16_t* keyId)
+{
+    uint16_t group;
+    uint16_t action;
+    uint16_t size;
+    int ret;
+    whPacket packet[1] = {0};
+    if (c == NULL || keyId == NULL)
+        return BAD_FUNC_ARG;
+    ret = wh_Client_RecvResponse(c, &group, &action, &size, (uint8_t*)packet);
+    if (ret == 0) {
+        if (packet->rc != 0)
+            ret = packet->rc;
+        else
+            *keyId = packet->keyCacheRes.id;
+    }
+    return ret;
+}
+
+int wh_Client_KeyEvictRequest(whClientContext* c, uint16_t keyId)
+{
+    whPacket packet[1] = {0};
+    if (c == NULL || keyId == WOLFHSM_ID_ERASED)
+        return BAD_FUNC_ARG;
+    /* set the keyId */
+    packet->keyEvictReq.id = keyId;
+    /* write request */
+    return wh_Client_SendRequest(c, WH_MESSAGE_GROUP_KEY, WH_KEY_EVICT,
+            WOLFHSM_PACKET_STUB_SIZE + sizeof(packet->keyEvictReq),
+            (uint8_t*)packet);
+}
+
+int wh_Client_KeyEvictResponse(whClientContext* c)
+{
+    uint16_t group;
+    uint16_t action;
+    uint16_t size;
+    int ret;
+    whPacket packet[1] = {0};
+    if (c == NULL)
+        return BAD_FUNC_ARG;
+    size = sizeof(packet);
+    ret = wh_Client_RecvResponse(c, &group, &action, &size, (uint8_t*)packet);
+    if (ret == 0) {
+        if (packet->rc != 0)
+            ret = packet->rc;
+    }
+    return ret;
+}
+
+int wh_Client_KeyExportRequest(whClientContext* c, uint16_t keyId)
+{
+    whPacket packet[1] = {0};
+    if (c == NULL || keyId == WOLFHSM_ID_ERASED)
+        return BAD_FUNC_ARG;
+    /* set keyId */
+    packet->keyExportReq.id = keyId;
+    /* write request */
+    return wh_Client_SendRequest(c, WH_MESSAGE_GROUP_KEY, WH_KEY_EXPORT,
+            WOLFHSM_PACKET_STUB_SIZE + sizeof(packet->keyExportReq),
+            (uint8_t*)packet);
+}
+
+int wh_Client_KeyExportResponse(whClientContext* c, uint8_t* label,
+    uint8_t* out, uint32_t* outSz)
+{
+    uint16_t group;
+    uint16_t action;
+    uint16_t size;
+    int ret;
+    uint8_t rawPacket[WH_COMM_MTU] = {0};
+    whPacket* packet = (whPacket*)rawPacket;
+    uint8_t* packOut = (uint8_t*)(&packet->keyExportRes + 1);
+    if (c == NULL || label == NULL || outSz == NULL)
+        return BAD_FUNC_ARG;
+    size = sizeof(rawPacket);
+    ret = wh_Client_RecvResponse(c, &group, &action, &size, rawPacket);
+    if (ret == 0) {
+        if (packet->rc != 0)
+            ret = packet->rc;
+        else  {
+            if (out == NULL) {
+                *outSz = packet->keyExportRes.len;
+                ret = LENGTH_ONLY_E;
```

Comment:
I don't get why we shouldn't use wolfssl error codes, removing all the wolfCrypt source code for a no crypto build makes sense but re-defining all these errors that already exist in wolfssl doesn't. Wolfssl is permanently coupled to wolfHSM and even if the source code is disabled the headers will not be.

Reply:
There's definitely an argument for that. My thoughts are that while wolfHSM might be closely coupled to wolfSSL internally (though not always, given the no crypto option), I don't see it as always being that way from the API perspective. I think that needing to look through source/docs for a code base that is different than the one you are using just to find the definition of a return code from the API function you just called is clunky from a user perspective. The return codes for the API functions wouldn't even be defined in our code base, for example.  It is one thing if, internally, we handle wolfCrypt return codes or pass them around, but returning wolfSSL codes to external users via API functions is a different story. I think in the small number of cases where a wolfHSM API function fails because of an error in wolfCrypt, we should return the category of failure via wolfHSM error code from the API function and provide the wolfCrypt error as an output argument (similar to how the client functions propagate the `server_rc`). I don't need to die on this hill at all, and am curious what @billphipps thinks, but my gut feel is that wolfHSM API functions should only return wolfHSM error codes for simplicity and clear separation of concerns. It is a small thing we can do to preserve simplicity to the user. Definitely worth a discussion though!

Reply:
Concur with Brett.  If it is a wolfHSM library call, we want to return simple, actionable results without extraneous details because we are in an embedded system and there is not troubleshooting possible.  If it is a wolfCrypt call that happens to be called via wolfHSM, then wolfCrypt errors should be used, but any actual failure will likely be fatal.  Good example here is the HW cryptocb's that are port specific.  Init and cleanup calls need to return wolfHSM errors while anything that comes out of the cryptocb must return wolfCrypt errors.  We squash the port-specific error returns to either wolfCrypt or wolfHSM errors without remorse.

Duplication of things like BAD_FUNC_ARG with WH_ERROR_BADARGS is expected.  wolfHSM basically has 4 different returns:
1. Action requested was successful and complete: WH_ERROR_OK
2. Action was NOT performed due to invalid input parameters or invalid state: WH_ERROR_BADARGS
3. Action was cannot be completed due to a fatal error and there is no reason to retry: WH_ERROR_ABORTED
4. Action requested was started but not completed OR a previous request has not completed yet.  The call should be retried: WH_ERROR_NOTREADY.

The other wolfHSM returns are used to indicate results that are not necessarily expected but not an error either (like NOTVERIFIED and NOTBLANK).  Both indicate the request completed but the result may not be what was expected.